compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConvertToDma.cpp

```diff
@@ -146,129 +154,400 @@ LogicalResult unPackDmaInputs(IREE::LinalgExt::UnPackOp unPackOp,
   return success();
 }
 
-/// Examines an input/output of a pack/unpack op and provides the
-/// corresponding offsets, sizes and strides required by the dma op
-LogicalResult setDmaInputs(Operation *&operandOp,
-                           SmallVector<OpFoldResult> &offsets,
+static bool isAllocation(Operation *op) {
+  return op && (isa<memref::AllocOp>(op) ||
+                isa<IREE::HAL::InterfaceBindingSubspanOp>(op));
+}
+
+/// Initialize offsets, sizes, and strides from an allocation operation.
+LogicalResult setFromAlloc(Operation *op, SmallVector<OpFoldResult> &offsets,
                            SmallVector<OpFoldResult> &sizes,
                            SmallVector<OpFoldResult> &strides) {
-  MLIRContext *ctx = operandOp->getContext();
-  if (auto allocOp = dyn_cast<memref::AllocOp>(operandOp)) {
-    auto [stridesI64, baseOffset] = getStridesAndOffset(allocOp.getType());
-    if (baseOffset != 0) {
-      auto message = llvm::formatv(
-          "with non-zero base offset {0} is not supported by the "
-          "current pass, requires testing and possible code changes.",
-          baseOffset);
-      return allocOp->emitOpError(message);
+  assert(isAllocation(op) &&
+         "expected memref.alloc or hal.interface.binding.subspan ");
+
+  MemRefType memRefType = cast<MemRefType>(op->getResult(0).getType());
+  MLIRContext *ctx = memRefType.getContext();
+  auto [stridesI64, baseOffset] = getStridesAndOffset(memRefType);
+  strides = getAsIndexOpFoldResult(ctx, stridesI64);
+
+  if (baseOffset != 0) {
+    auto message = llvm::formatv(
+        "has non-zero offset {0} which is currently unsupported.", baseOffset);
+    return op->emitOpError(message);
+  }
+  offsets.resize(strides.size(), getAsIndexOpFoldResult(ctx, 0));
+
+  ArrayRef<int64_t> sizesI64 = memRefType.getShape();
+  if (llvm::any_of(sizesI64,
+                   [](int64_t size) { return ShapedType::isDynamic(size); })) {
+    return op->emitOpError("has dynamic size, which is unsupported in DMA.");
+  }
+  sizes = getAsIndexOpFoldResult(ctx, sizesI64);
+  return success();
+}
+
+/// Return a + b.
+SmallVector<OpFoldResult> getIndexOpFoldResultSum(OpBuilder &builder,
+                                                  Location loc,
+                                                  ArrayRef<OpFoldResult> lhs,
+                                                  ArrayRef<OpFoldResult> rhs) {
+  assert(lhs.size() == rhs.size() && "a and b not same size");
+  SmallVector<OpFoldResult> sum;
+  sum.reserve(lhs.size());
+
+  auto getConstant = [&](int64_t v) -> Value {
+    return builder.create<arith::ConstantOp>(
+        loc, builder.getIndexType(),
+        builder.getIntegerAttr(builder.getIndexType(), v));
+  };
+
+  auto add = [&](Value v, IntegerAttr attr) {
+    if (attr.getInt() == 0) return v;
+    return builder.create<arith::AddIOp>(loc, v, getConstant(attr.getInt()))
```

Comment:
I'm wondering if there's an easier way / existing utility in upstream that can do this, instead of creating new ops.

Reply:
Yeah, I wondered the same thing.